exercise/exercise_models.py

```diff
@@ -706,13 +707,22 @@ def grade(self, request, submission, no_penalties=False, url_name="exercise"):
             submission.ordinal_number(), url_name, submission_url
         )
         try:
-            return load_feedback_page(
-                request, url, self, submission, no_penalties=no_penalties
-            )
+            if use_messages:
+                return load_feedback_page(
```

Comment:
I will try to do this.

Reply:
I changed it so that the `grade` and `load_feedback_page` never use the messages framework. The normal web view classes set messages after calling `grade`, while the API can add the error messages to the JSON response.